source/module_psi/include/gpu_cuda.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include <vector>
+#include <stdio.h>
+#include <assert.h>
+#include <cuda_runtime.h>
+
+namespace gpu_cuda {
+
+template <typename FPTYPE>
+void memcpy_device_to_host(
+    const FPTYPE * device, 
+    std::vector<FPTYPE> &host) 
+{
+  cudaMemcpy(&host[0], device, sizeof(FPTYPE) * host.size(), cudaMemcpyDeviceToHost);  
+}
+
+template <typename FPTYPE>
+void malloc_host_memory_sync(
+    FPTYPE * &host,
+    const FPTYPE * device,
+    const int size)
+{
+  host = (FPTYPE *)malloc(sizeof(FPTYPE) * size);
+  memcpy_device_to_host(device, host, size);
+}
+
+template <typename FPTYPE>
+void memcpy_host_to_device(
+    FPTYPE * device, 
+    const FPTYPE * host,
+    const int size) 
+{
+  cudaMemcpy(device, host, sizeof(FPTYPE) * size, cudaMemcpyHostToDevice);  
+}
+
+template <typename FPTYPE>
+void malloc_device_memory_sync(
+    FPTYPE * &device,
+    const FPTYPE * host,
+    const int size)
+{
+  cudaMalloc((void **)&device, sizeof(FPTYPE) * size);
+  memcpy_host_to_device(device, host, size);
+}
+
+template <typename FPTYPE>
+void malloc_device_memory(
+    FPTYPE * &device,
+    const int size)
+{
+  cudaMalloc((void **)&device, sizeof(FPTYPE) * size);
+}
+
+template <typename FPTYPE>
+void delete_device_memory(
+    FPTYPE * &device) 
+{
+  if (device != NULL) {
+    cudaFree(device);
+  }
+}
+
+template <typename FPTYPE>
+void memset_device_memory(
+    FPTYPE * device, 
+    const int var,
+    const int size) 
+{
+  cudaMemset(device, var, sizeof(FPTYPE) * size);  
+}
+
+template <typename FPTYPE>
+void abacus_resize_memory(
```

Comment:
This method seems like to be `realloc`. However, it would delete all the data. I suggest removing this method.

Reply:
I have changed the name from resize to reallocate. @dyzheng @caic99  Is it acceptable to change this line from resize to reallocate? Does psi need to keep the original value ?
https://github.com/deepmodeling/abacus-develop/blob/85341c33cf1f286b1743bed92719b810bbc5b2e8/source/module_psi/psi.h#L128

Reply:
> realloc

@denghuilu I think `reallocate` might mislead the developers more. I mean, `realloc` preserves data, but this method won't. Personally, I suggest using a delete and an alloc, instead of combining them into one function. WDYT?

Reply:
> Does psi need to keep the original value ?

No. @dyzheng am I right?

Reply:
Since the class method is resize, I suggest to keep the resize function name. And add a description(code comment) for this function call to avoid the misleading to the other developers.